File: backend/src/jobs/registry.rs
use std::collections::HashMap;
use std::sync::{Arc, Mutex};

use uuid::Uuid;

use crate::api::{JobKind, JobProgress, JobResult, JobSnapshot, JobState};
use crate::error::{BackendError, BackendResult};

#[derive(Debug, Clone)]
pub enum StartJobOutcome {
    Created(JobSnapshot),
    Existing(JobSnapshot),
}

#[derive(Debug, Clone, Default)]
pub struct JobRegistry {
    inner: Arc<Mutex<JobRegistryInner>>,
}

#[derive(Debug, Default)]
struct JobRegistryInner {
    jobs: HashMap<String, JobEntry>,
    active_fingerprints: HashMap<String, String>,
}

#[derive(Debug, Clone)]
struct JobEntry {
    fingerprint: Option<String>,
    cancellation_requested: bool,
    snapshot: JobSnapshot,
}

impl JobRegistry {
    pub fn start_job(
        &self,
        job_kind: JobKind,
        study_id: Option<String>,
        fingerprint: String,
    ) -> BackendResult<StartJobOutcome> {
        let mut inner = self.lock()?;

        if let Some(existing_job_id) = inner.active_fingerprints.get(&fingerprint)
            && let Some(existing) = inner.jobs.get(existing_job_id)
        {
            return Ok(StartJobOutcome::Existing(existing.snapshot.clone()));
        }

        let job_id = Uuid::new_v4().to_string();
        let snapshot = JobSnapshot {
            job_id: job_id.clone(),
            job_kind,
            study_id,
            state: JobState::Queued,
            progress: JobProgress {
                percent: 0,
                stage: String::from("queued"),
                message: String::from("Queued"),
            },
            from_cache: false,
            result: None,
            error: None,
        };
        inner.active_fingerprints.insert(fingerprint.clone(), job_id.clone());
        inner.jobs.insert(
            job_id,
            JobEntry {
                fingerprint: Some(fingerprint),
                cancellation_requested: false,
                snapshot: snapshot.clone(),
            },
        );

        Ok(StartJobOutcome::Created(snapshot))
    }

    pub fn create_cached_job(
        &self,
        job_kind: JobKind,
        study_id: Option<String>,
        result: JobResult,
    ) -> BackendResult<JobSnapshot> {
        let mut inner = self.lock()?;
        let job_id = Uuid::new_v4().to_string();
        let snapshot = JobSnapshot {
            job_id: job_id.clone(),
            job_kind,
            study_id,
            state: JobState::Completed,
            progress: JobProgress {
                percent: 100,
                stage: String::from("cacheHit"),
                message: String::from("Loaded from cache"),
            },
            from_cache: true,
            result: Some(result),
            error: None,
        };
        inner.jobs.insert(
            job_id,
            JobEntry {
                fingerprint: None,
                cancellation_requested: false,
                snapshot: snapshot.clone(),
            },
        );
        Ok(snapshot)
    }

    pub fn get(&self, job_id: &str) -> BackendResult<JobSnapshot> {
        let inner = self.lock()?;
        inner
            .jobs
            .get(job_id)
            .map(|entry| entry.snapshot.clone())
            .ok_or_else(|| BackendError::not_found(format!("job not found: {job_id}")))
    }

    pub fn update_progress(
        &self,
        job_id: &str,
        state: JobState,
        percent: u8,
        stage: impl Into<String>,
        message: impl Into<String>,
    ) -> BackendResult<JobSnapshot> {
        let mut inner = self.lock()?;
        let entry = inner
            .jobs
            .get_mut(job_id)
            .ok_or_else(|| BackendError::not_found(format!("job not found: {job_id}")))?;
        if matches!(
            entry.snapshot.state,
            JobState::Completed | JobState::Failed | JobState::Cancelled
        ) {
            return Ok(entry.snapshot.clone());
        }
        entry.snapshot.state = state;
        entry.snapshot.progress = JobProgress {
            percent,
            stage: stage.into(),
            message: message.into(),
        };
        entry.snapshot.error = None;
        Ok(entry.snapshot.clone())
    }

    pub fn complete(&self, job_id: &str, result: JobResult) -> BackendResult<JobSnapshot> {
        let mut inner = self.lock()?;
        let fingerprint = {
            let entry = inner
                .jobs
                .get_mut(job_id)
                .ok_or_else(|| BackendError::not_found(format!("job not found: {job_id}")))?;
            entry.snapshot.state = JobState::Completed;
            entry.snapshot.progress = JobProgress {
                percent: 100,
                stage: String::from("completed"),
                message: String::from("Completed"),
            };
            entry.snapshot.result = Some(result);
            entry.snapshot.error = None;
            entry.fingerprint.take()
        };

        if let Some(fingerprint) = fingerprint {
            inner.active_fingerprints.remove(&fingerprint);
        }

        inner
            .jobs
            .get(job_id)
            .map(|entry| entry.snapshot.clone())
            .ok_or_else(|| BackendError::not_found(format!("job not found: {job_id}")))
    }

    pub fn fail(&self, job_id: &str, error: BackendError) -> BackendResult<JobSnapshot> {
        let mut inner = self.lock()?;
        let fingerprint = {
            let entry = inner
                .jobs
                .get_mut(job_id)
                .ok_or_else(|| BackendError::not_found(format!("job not found: {job_id}")))?;
            entry.snapshot.state = JobState::Failed;
            entry.snapshot.progress.message = String::from("Failed");
            entry.snapshot.error = Some(error);
            entry.fingerprint.take()
        };

        if let Some(fingerprint) = fingerprint {
            inner.active_fingerprints.remove(&fingerprint);
        }

        inner
            .jobs
            .get(job_id)
            .map(|entry| entry.snapshot.clone())
            .ok_or_else(|| BackendError::not_found(format!("job not found: {job_id}")))
    }

    pub fn cancel(&self, job_id: &str) -> BackendResult<JobSnapshot> {
        let mut inner = self.lock()?;
        let mut fingerprint_to_remove = None;
        let snapshot = {
            let entry = inner
                .jobs
                .get_mut(job_id)
                .ok_or_else(|| BackendError::not_found(format!("job not found: {job_id}")))?;
            match entry.snapshot.state {
                JobState::Queued => {
                    entry.cancellation_requested = true;
                    entry.snapshot.state = JobState::Cancelled;
                    entry.snapshot.progress.message = String::from("Cancelled before start");
                    fingerprint_to_remove = entry.fingerprint.take();
                }
                JobState::Running | JobState::Cancelling => {
                    entry.cancellation_requested = true;
                    entry.snapshot.state = JobState::Cancelling;
                    entry.snapshot.progress.message = String::from("Cancellation requested");
                }
                JobState::Completed | JobState::Failed | JobState::Cancelled => {}
            }
            entry.snapshot.clone()
        };

        if let Some(fingerprint) = fingerprint_to_remove {
            inner.active_fingerprints.remove(&fingerprint);
        }

        Ok(snapshot)
    }

    pub fn mark_cancelled(
        &self,
        job_id: &str,
        stage: impl Into<String>,
        message: impl Into<String>,
    ) -> BackendResult<JobSnapshot> {
        let mut inner = self.lock()?;
        let fingerprint = {
            let entry = inner
                .jobs
                .get_mut(job_id)
                .ok_or_else(|| BackendError::not_found(format!("job not found: {job_id}")))?;
            entry.snapshot.state = JobState::Cancelled;
            entry.snapshot.progress.stage = stage.into();
            entry.snapshot.progress.message = message.into();
            entry.snapshot.error = None;
            entry.fingerprint.take()
        };

        if let Some(fingerprint) = fingerprint {
            inner.active_fingerprints.remove(&fingerprint);
        }

        inner
            .jobs
            .get(job_id)
            .map(|entry| entry.snapshot.clone())
            .ok_or_else(|| BackendError::not_found(format!("job not found: {job_id}")))
    }

    pub fn is_cancellation_requested(&self, job_id: &str) -> BackendResult<bool> {
        let inner = self.lock()?;
        inner
            .jobs
            .get(job_id)
            .map(|entry| entry.cancellation_requested)
            .ok_or_else(|| BackendError::not_found(format!("job not found: {job_id}")))
    }

    fn lock(&self) -> BackendResult<std::sync::MutexGuard<'_, JobRegistryInner>> {
        self.inner
            .lock()
            .map_err(|_| BackendError::internal("job registry is unavailable"))
    }
}

#[cfg(test)]
mod tests {
    use crate::api::{JobKind, ProcessStudyCommandResult};

    use super::{JobRegistry, StartJobOutcome};

    #[test]
    fn start_job_reuses_active_duplicate_requests() {
        let registry = JobRegistry::default();

        let first = registry
            .start_job(
                JobKind::ProcessStudy,
                Some(String::from("study-1")),
                String::from("fingerprint-1"),
            )
            .expect("start first job");
        let second = registry
            .start_job(
                JobKind::ProcessStudy,
                Some(String::from("study-1")),
                String::from("fingerprint-1"),
            )
            .expect("start duplicate job");

        let StartJobOutcome::Created(first_snapshot) = first else {
            panic!("expected first job to be created");
        };
        let StartJobOutcome::Existing(second_snapshot) = second else {
            panic!("expected duplicate job to be reused");
        };

        assert_eq!(first_snapshot.job_id, second_snapshot.job_id);
    }

    #[test]
    fn complete_job_releases_fingerprint_for_later_runs() {
        let registry = JobRegistry::default();
        let created = registry
            .start_job(
                JobKind::ProcessStudy,
                Some(String::from("study-1")),
                String::from("fingerprint-1"),
            )
            .expect("start job");

        let StartJobOutcome::Created(snapshot) = created else {
            panic!("expected a created job");
        };

        registry
            .complete(
                &snapshot.job_id,
                crate::api::JobResult::ProcessStudy(ProcessStudyCommandResult {
                    study_id: String::from("study-1"),
                    preview_path: "/tmp/preview.png".into(),
                    dicom_path: "/tmp/output.dcm".into(),
                    loaded_width: 1,
                    loaded_height: 1,
                    mode: String::from("grayscale"),
                    measurement_scale: None,
                }),
            )
            .expect("complete job");

        let next = registry
            .start_job(
                JobKind::ProcessStudy,
                Some(String::from("study-1")),
                String::from("fingerprint-1"),
            )
            .expect("start replacement job");

        assert!(matches!(next, StartJobOutcome::Created(_)));
    }
}

File: go-backend/internal/jobs/registry.go
package jobs

import (
	"context"
	"fmt"
	"sync"

	"xrayview/go-backend/internal/contracts"
)

type StartJobOutcome struct {
	Snapshot contracts.JobSnapshot
	Created  bool
}

type Registry struct {
	mu                 sync.RWMutex
	newJobID           idGenerator
	jobs               map[string]*registryEntry
	activeFingerprints map[string]string
}

type registryEntry struct {
	fingerprint           string
	cancellationRequested bool
	snapshot              contracts.JobSnapshot
	cancel                context.CancelFunc
}

func NewRegistry(jobIDFactory idGenerator) *Registry {
	if jobIDFactory == nil {
		jobIDFactory = generateJobID
	}

	return &Registry{
		newJobID:           jobIDFactory,
		jobs:               make(map[string]*registryEntry),
		activeFingerprints: make(map[string]string),
	}
}

func (registry *Registry) StartJob(
	jobKind contracts.JobKind,
	studyID string,
	fingerprint string,
) (StartJobOutcome, error) {
	registry.mu.Lock()
	defer registry.mu.Unlock()

	if existingJobID, ok := registry.activeFingerprints[fingerprint]; ok {
		if existing, exists := registry.jobs[existingJobID]; exists &&
			!isTerminalState(existing.snapshot.State) {
			return StartJobOutcome{
				Snapshot: cloneJobSnapshot(existing.snapshot),
				Created:  false,
			}, nil
		}

		delete(registry.activeFingerprints, fingerprint)
	}

	jobID, err := registry.newJobID()
	if err != nil {
		return StartJobOutcome{}, contracts.Internal(fmt.Sprintf("generate job id: %v", err))
	}

	snapshot := queuedJobSnapshot(jobID, jobKind, studyID)
	registry.jobs[jobID] = &registryEntry{
		fingerprint: fingerprint,
		snapshot:    snapshot,
	}
	registry.activeFingerprints[fingerprint] = jobID

	return StartJobOutcome{
		Snapshot: cloneJobSnapshot(snapshot),
		Created:  true,
	}, nil
}

func (registry *Registry) AttachCancel(jobID string, cancel context.CancelFunc) error {
	registry.mu.Lock()
	defer registry.mu.Unlock()

	entry, ok := registry.jobs[jobID]
	if !ok {
		return contracts.NotFound(fmt.Sprintf("job not found: %s", jobID))
	}

	entry.cancel = cancel
	if entry.cancellationRequested && cancel != nil {
		cancel()
	}

	return nil
}

func (registry *Registry) CreateCachedJob(
	jobKind contracts.JobKind,
	studyID string,
	result contracts.JobResult,
) (contracts.JobSnapshot, error) {
	registry.mu.Lock()
	defer registry.mu.Unlock()

	jobID, err := registry.newJobID()
	if err != nil {
		return contracts.JobSnapshot{}, contracts.Internal(fmt.Sprintf("generate job id: %v", err))
	}

	snapshot := completedJobSnapshot(jobID, jobKind, studyID, true, result)
	registry.jobs[jobID] = &registryEntry{snapshot: snapshot}

	return cloneJobSnapshot(snapshot), nil
}

func (registry *Registry) Get(jobID string) (contracts.JobSnapshot, error) {
	registry.mu.RLock()
	defer registry.mu.RUnlock()

	entry, ok := registry.jobs[jobID]
	if !ok {
		return contracts.JobSnapshot{}, contracts.NotFound(fmt.Sprintf("job not found: %s", jobID))
	}

	return cloneJobSnapshot(entry.snapshot), nil
}

func (registry *Registry) UpdateProgress(
	jobID string,
	state contracts.JobState,
	percent int,
	stage string,
	message string,
) (contracts.JobSnapshot, error) {
	registry.mu.Lock()
	defer registry.mu.Unlock()

	entry, ok := registry.jobs[jobID]
	if !ok {
		return contracts.JobSnapshot{}, contracts.NotFound(fmt.Sprintf("job not found: %s", jobID))
	}
	if isTerminalState(entry.snapshot.State) || entry.snapshot.State == contracts.JobStateCancelling {
		return cloneJobSnapshot(entry.snapshot), nil
	}

	entry.snapshot.State = state
	entry.snapshot.Progress = contracts.JobProgress{
		Percent: percent,
		Stage:   stage,
		Message: message,
	}
	entry.snapshot.Error = nil

	return cloneJobSnapshot(entry.snapshot), nil
}

func (registry *Registry) Complete(
	jobID string,
	result contracts.JobResult,
) (contracts.JobSnapshot, error) {
	registry.mu.Lock()
	defer registry.mu.Unlock()

	entry, ok := registry.jobs[jobID]
	if !ok {
		return contracts.JobSnapshot{}, contracts.NotFound(fmt.Sprintf("job not found: %s", jobID))
	}

	if registry.shouldCancelLocked(entry) {
		registry.markCancelledLocked(entry, entry.snapshot.Progress.Stage, "Cancelled by user")
		return cloneJobSnapshot(entry.snapshot), nil
	}
	if isTerminalState(entry.snapshot.State) {
		return cloneJobSnapshot(entry.snapshot), nil
	}

	entry.snapshot.State = contracts.JobStateCompleted
	entry.snapshot.Progress = contracts.JobProgress{
		Percent: 100,
		Stage:   "completed",
		Message: "Completed",
	}
	entry.snapshot.FromCache = false
	entry.snapshot.Result = &contracts.JobResult{
		Kind:    result.Kind,
		Payload: result.Payload,
	}
	entry.snapshot.Error = nil
	entry.cancellationRequested = false
	entry.cancel = nil
	registry.releaseFingerprintLocked(entry)

	return cloneJobSnapshot(entry.snapshot), nil
}

func (registry *Registry) Fail(jobID string, err error) (contracts.JobSnapshot, error) {
	backendErr, ok := err.(contracts.BackendError)
	if !ok {
		backendErr = contracts.Internal(err.Error())
	}

	registry.mu.Lock()
	defer registry.mu.Unlock()

	entry, ok := registry.jobs[jobID]
	if !ok {
		return contracts.JobSnapshot{}, contracts.NotFound(fmt.Sprintf("job not found: %s", jobID))
	}

	if registry.shouldCancelLocked(entry) {
		registry.markCancelledLocked(entry, entry.snapshot.Progress.Stage, "Cancelled by user")
		return cloneJobSnapshot(entry.snapshot), nil
	}
	if isTerminalState(entry.snapshot.State) {
		return cloneJobSnapshot(entry.snapshot), nil
	}

	entry.snapshot.State = contracts.JobStateFailed
	entry.snapshot.Progress.Message = "Failed"
	entry.snapshot.Result = nil
	entry.snapshot.Error = &backendErr
	entry.cancel = nil
	registry.releaseFingerprintLocked(entry)

	return cloneJobSnapshot(entry.snapshot), nil
}

func (registry *Registry) Cancel(jobID string) (contracts.JobSnapshot, error) {
	registry.mu.Lock()
	defer registry.mu.Unlock()

	entry, ok := registry.jobs[jobID]
	if !ok {
		return contracts.JobSnapshot{}, contracts.NotFound(fmt.Sprintf("job not found: %s", jobID))
	}

	switch entry.snapshot.State {
	case contracts.JobStateQueued:
		entry.cancellationRequested = true
		entry.snapshot.State = contracts.JobStateCancelled
		entry.snapshot.Progress.Message = "Cancelled before start"
		entry.snapshot.Result = nil
		entry.snapshot.Error = nil
		if entry.cancel != nil {
			entry.cancel()
		}
		entry.cancel = nil
		registry.releaseFingerprintLocked(entry)
	case contracts.JobStateRunning, contracts.JobStateCancelling:
		entry.cancellationRequested = true
		entry.snapshot.State = contracts.JobStateCancelling
		entry.snapshot.Progress.Message = "Cancellation requested"
		entry.snapshot.Error = nil
		if entry.cancel != nil {
			entry.cancel()
		}
	case contracts.JobStateCompleted, contracts.JobStateFailed, contracts.JobStateCancelled:
	}

	return cloneJobSnapshot(entry.snapshot), nil
}

func (registry *Registry) MarkCancelled(
	jobID string,
	stage string,
	message string,
) (contracts.JobSnapshot, error) {
	registry.mu.Lock()
	defer registry.mu.Unlock()

	entry, ok := registry.jobs[jobID]
	if !ok {
		return contracts.JobSnapshot{}, contracts.NotFound(fmt.Sprintf("job not found: %s", jobID))
	}
	if entry.snapshot.State == contracts.JobStateCompleted || entry.snapshot.State == contracts.JobStateFailed {
		return cloneJobSnapshot(entry.snapshot), nil
	}

	registry.markCancelledLocked(entry, stage, message)
	return cloneJobSnapshot(entry.snapshot), nil
}

func (registry *Registry) IsCancellationRequested(jobID string) (bool, error) {
	registry.mu.RLock()
	defer registry.mu.RUnlock()

	entry, ok := registry.jobs[jobID]
	if !ok {
		return false, contracts.NotFound(fmt.Sprintf("job not found: %s", jobID))
	}

	return entry.cancellationRequested, nil
}

func (registry *Registry) shouldCancelLocked(entry *registryEntry) bool {
	return entry.cancellationRequested ||
		entry.snapshot.State == contracts.JobStateCancelling ||
		entry.snapshot.State == contracts.JobStateCancelled
}

func (registry *Registry) markCancelledLocked(
	entry *registryEntry,
	stage string,
	message string,
) {
	entry.cancellationRequested = true
	entry.snapshot.State = contracts.JobStateCancelled
	if stage != "" {
		entry.snapshot.Progress.Stage = stage
	}
	if message != "" {
		entry.snapshot.Progress.Message = message
	}
	entry.snapshot.Result = nil
	entry.snapshot.Error = nil
	entry.cancel = nil
	registry.releaseFingerprintLocked(entry)
}

func (registry *Registry) releaseFingerprintLocked(entry *registryEntry) {
	if entry.fingerprint == "" {
		return
	}

	delete(registry.activeFingerprints, entry.fingerprint)
	entry.fingerprint = ""
}

func cloneJobSnapshot(snapshot contracts.JobSnapshot) contracts.JobSnapshot {
	cloned := snapshot

	if snapshot.StudyID != nil {
		cloned.StudyID = studyIDPointer(*snapshot.StudyID)
	}
	if snapshot.Result != nil {
		result := *snapshot.Result
		cloned.Result = &result
	}
	if snapshot.Error != nil {
		backendErr := *snapshot.Error
		if backendErr.Details != nil {
			backendErr.Details = append([]string(nil), backendErr.Details...)
		}
		cloned.Error = &backendErr
	}

	return cloned
}
